src/ripple/nodestore/Database.h

```diff
@@ -135,20 +135,23 @@ class Database : public Stoppable
         If I/O is required to determine whether or not the object is present,
         `false` is returned. Otherwise, `true` is returned and `object` is set
         to refer to the object, or `nullptr` if the object is not present.
-        If I/O is required, the I/O is scheduled.
+        If I/O is required, the I/O is scheduled and `true` is returned
 
         @note This can be called concurrently.
         @param hash The key of the object to retrieve
         @param ledgerSeq The sequence of the ledger where the
                 object is stored, used by the shard store.
         @param nodeObject The object retrieved
-        @return Whether the operation completed
+        @param callback Callback function when read completes
+        @return true if the operation completed, false if pending
     */
     virtual bool
     asyncFetch(
         uint256 const& hash,
         std::uint32_t ledgerSeq,
-        std::shared_ptr<NodeObject>& nodeObject) = 0;
+        std::shared_ptr<NodeObject>& nodeObject,
+        std::function<void(
+            std::shared_ptr<NodeObject>&)>&& callback) = 0;
```

Comment:
I'd like to refactor this interface. `nodeObject` is never used, we always get the `nodeObject` through the callback. All implementations always post a read, _except_ `shards`, which first checks that the ledger is part  of its collection. If it's not, it doesn't post a read, and will not retrieve the object. Given that, I'd like to get rid of the `nodeObject` parameter, and make the return value `true` if pending (i.e. "success") and `false` if the object will never be found (failure).

Here's a patch for that refactor: https://github.com/seelabs/rippled/commit/f272b74741a8bc9d0e51f279fa978f9ed4a8ecf9 

Two notes: 
* this patch is based on top of a clang-reformat patch. You need the reformat-patch before applying this patch).
* this patch contains a fix for `DatabaseShardImp::asyncFetch` that I describe in another comment

Reply:
It looks like `DatabaseShardImp::asyncFetch` can be reomved completely. In that case, this function can now return `void`. Here's a second patch that builds on the first: https://github.com/ripple/rippled/commit/6147739d1a71b522597b632988fb055170c707dc